server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -482,14 +484,18 @@ private void startRestore(
         ).flatMap(Collection::stream).distinct().toList();
 
         final Set<String> explicitlyRequestedSystemIndices = new HashSet<>();
+
+        @FixForMultiProject
+        final ProjectId projectId = ProjectId.DEFAULT;
+        ProjectMetadata.Builder projectBuilder = metadataBuilder.getProject(projectId);
```

Comment:
Looks like when `if (request.includeGlobalState())` is `false`, we do `metadataBuilder = Metadata.builder()`, which in turn causes an NPE here. (I derived that by looking at the failing tests, yay for tests)

I had a quick look at your other changes, but they seem to be safe from this scenario.